Add agent memory search for OpenClaw and Hermes Agent using llama.cpp (#2438)
huggingface/hub-docs: docs/hub/agents-local.md

You can also run `openclaw onboard` interactively, select `custom-compatibility` with `openai`, and pass the same configuration.

### Local Memory Search for OpenClaw

> You can run local embedding models with Llama.cpp for your agent's memory search. To do so, make sure to have node-llama-cpp.

Reviewer: I don't use OpenClaw. Does this have to be the `node` version?

Author: not sure I understand your question :/

Reviewer: why not just use llama.cpp, instead of node-llama-cpp? Is this an OpenClaw requirement?

Author: if you install this way you can use below one-liner to get llama server up and using it for memory instead of setting it up yourself

```bash
npm ls -g node-llama-cpp --depth=0

> ```

Reviewer: Should we expect it to be installed? What if it's not?

Author: it will not work, is this unclear?

Reviewer: My question is: instead of listing the package, should we just show an install command instead? Or is node-llama-cpp expected to be available in every OpenClaw installation?

Author: ack sorry I missed this

Author: addressed

Here's an example snippet to run [quantized EmbeddingGemma-300M](https://huggingface.co/ggml-org/embeddinggemma-300M-GGUF?show_file_info=embeddinggemma-300M-Q8_0.gguf) locally for memory search.

```bash
openclaw config set agents.defaults.memorySearch.provider local

> openclaw config set agents.defaults.memorySearch.local.modelPath "hf:ggml-org/embeddinggemma-300m-qat-q8_0-GGUF/embeddinggemma-300m-qat-Q8_0.gguf"

Reviewer: And does it know how to run the model just based on the model name?

Author: addressed this comment above

You can also run `openclaw onboard` interactively, select `custom-compatibility` with `openai`, and pass the same configuration.

### Local Memory Search for OpenClaw

You can run local embedding models with Llama.cpp for your agent's memory search. To do so, make sure to have node-llama-cpp. 

```bash
npm i node-llama-cpp 
```

> Here's an example snippet to run [quantized EmbeddingGemma-300M](https://huggingface.co/ggml-org/embeddinggemma-300M-GGUF?show_file_info=embeddinggemma-300M-Q8_0.gguf) locally for memory search. OpenClaw automatically downloads and serves the model with below command.

Reviewer: ```suggestion
Here's an example snippet to run [quantized EmbeddingGemma-300M](https://huggingface.co/ggml-org/embeddinggemma-300M-GGUF?show_file_info=embeddinggemma-300M-Q8_0.gguf) locally for memory search. OpenClaw automatically downloads and serves the model with the command below.
```

    model: ggml-org/gemma-4-26B-A4B-it-GGUF:Q4_K_M
```

### Local Memory Search for Hermes Agent

> Hermes Agent consumes served semantic search models through endpoints. Once you get the model up on endpoint 8080 with the inference engine of your choice, add the following to `~/.hermes/config.yaml`.

Reviewer: ```suggestion
Hermes Agent consumes semantic search models through endpoints. Once you get your preferred embedding model up on endpoint 8080 with llama.cpp or the inference engine of your choice, add the following to `~/.hermes/config.yaml`.
```

Perhaps we should show an example?

```bash
auxiliary:
  session_search:
    base_url: "http://127.0.0.1:8080/v1"
    api_key: "no-key-required"
    model: "local-llama" # your model alias
    timeout: 90
    max_concurrency: 1
```

> Check if this works, none - built-in only shows that no other memory plug-ins are used.

Reviewer: ```suggestion
Check if this works, `none - built-in only` shows that no other memory plug-ins are used.
```

```bash
$ hermes memory status
# Memory status
#────────────────────────────────────────
#  Built-in:  always active

> #  Provider:  (none — built-in only)

Reviewer: So this means it's working?

Author: I have addressed this comment

    model: ggml-org/gemma-4-26B-A4B-it-GGUF:Q4_K_M
```

### Local Memory Search for Hermes Agent

> Hermes Agent consumes semantic search models through endpoints. Once you get your preferred embedding model up on endpoint 8080 with llama.cpp or the inference engine of your choice, add the following to `~/.hermes/config.yaml`.

Reviewer: How does memory search work if you don't do anything? Because it's able to look up things in memory by default, AFAIU. Here's my output from the command you mention below. I haven't done anything about semantic search, but the first lines match what you said:

```
Memory status
────────────────────────────────────────
  Built-in:  always active
  Provider:  (none — built-in only)

  Installed plugins:
    • byterover  (requires API key)
    • hindsight  (API key / local)
    • holographic  (local)
    • honcho  (API key / local)
    • mem0  (API key / local)
    • openviking  (API key / local)
    • retaindb  (API key / local)
    • supermemory  (requires API key)
```

Author: it does exact text search afaik, where it just does look-up from sessions. imo it's counterproductive

```bash
auxiliary:
  session_search:
    base_url: "http://127.0.0.1:8080/v1"
    api_key: "no-key-required"
    model: "local-llama" # your model alias
    timeout: 90
    max_concurrency: 1
```

> Check if this works, `none - built-in only` shows that no other memory plug-ins are used. Below output shows that local serving is active.

Reviewer: ```suggestion
Check if this works, `none - built-in only` shows that no other memory plug-ins are used. The output below shows that local serving is active.
```
